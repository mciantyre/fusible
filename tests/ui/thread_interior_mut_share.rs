// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: Copyright 2024 Ian McIntyre

use core::{cell::Cell, pin::pin};
use fusible::thread::{StaticStack, Thread};

static STACK: StaticStack<512> = StaticStack::new();

/// Disallowed: Cell (or any !Sync type) cannot be shared across threads.
fn main() {
    let state = Cell::new(5);
    let thread = pin!(Thread::context());

    fusible::kernel_enter(|app_define| {
        app_define
            .create_thread(
                thread.into_ref(),
                STACK.take().unwrap(),
                &Default::default(),
                || state.set(state.get() + 4),
            )
            .unwrap();
    });
}
