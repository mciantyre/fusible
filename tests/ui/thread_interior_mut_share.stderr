error[E0277]: `Cell<i32>` cannot be shared between threads safely
  --> tests/ui/thread_interior_mut_share.rs:20:17
   |
16 |             .create_thread(
   |              ------------- required by a bound introduced by this call
...
20 |                 || state.set(state.get() + 4),
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Cell<i32>` cannot be shared between threads safely
   |
   = help: the trait `Sync` is not implemented for `Cell<i32>`
   = note: if you want to do aliasing and mutation between multiple threads, use `std::sync::RwLock` or `std::sync::atomic::AtomicI32` instead
   = note: required for `&Cell<i32>` to implement `Send`
note: required because it's used within this closure
  --> tests/ui/thread_interior_mut_share.rs:20:17
   |
20 |                 || state.set(state.get() + 4),
   |                 ^^
note: required by a bound in `fusible::thread::<impl AppDefine<'_, 'pke>>::create_thread`
  --> src/thread.rs
   |
   |     pub fn create_thread<'t, 'ctx, F>(
   |            ------------- required by a bound in this associated function
...
   |         F: FnOnce() + Send + 'pke,
   |                       ^^^^ required by this bound in `fusible::thread::<impl AppDefine<'_, 'pke>>::create_thread`
